File: Backend/mapping/src/test/java/com/mhp/coding/challenges/mapping/mappers/ArticleMapperTest.java
package com.mhp.coding.challenges.mapping.mappers;

import com.mhp.coding.challenges.mapping.models.db.Article;
import com.mhp.coding.challenges.mapping.models.db.Image;
import com.mhp.coding.challenges.mapping.models.db.ImageSize;
import com.mhp.coding.challenges.mapping.models.db.blocks.*;
import com.mhp.coding.challenges.mapping.models.dto.ArticleDto;
import com.mhp.coding.challenges.mapping.models.dto.ImageDto;
import com.mhp.coding.challenges.mapping.models.dto.blocks.*;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringRunner;

import java.time.Instant;
import java.util.Date;
import java.util.HashSet;

import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;

@RunWith(SpringRunner.class)
@ContextConfiguration(classes = ArticleMapperTest.SpringTestConfig.class)
public class ArticleMapperTest {
    @Configuration
    @ComponentScan(basePackageClasses = ArticleMapperTest.class)
    public static class SpringTestConfig {
    }

    @Autowired
    private ArticleMapper articleMapper;

    // acceptance criterion 1 - mapping
    // acceptance criterion 2 - sorted
    @Test
    public void shouldMapFullArticle() {
        // given
        final Date someDate = Date.from(Instant.EPOCH);
        Article article = Article.builder()
                .author("author")
                .id(12345678L)
                .description("description")
                .lastModified(someDate)
                .lastModifiedBy("lastModifiedBy")
                .title("title")
                .blocks(new HashSet<>(asList(
                        TextBlock.builder()
                                .sortIndex(3)
                                .text("text")
                                .build(),
                        GalleryBlock.builder()
                                .sortIndex(2)
                                .images(asList(
                                        Image.builder()
                                                .id(123L)
                                                .imageSize(ImageSize.LARGE)
                                                .url("url")
                                                .lastModified(someDate)
                                                .lastModifiedBy("lastModifiedBy")
                                                .build()
                                )).build(),
                        ImageBlock.builder()
                                .sortIndex(1)
                                .image(Image.builder()
                                        .id(123L)
                                        .imageSize(ImageSize.MEDIUM)
                                        .url("other_url")
                                        .lastModified(someDate)
                                        .lastModifiedBy("lastModifiedBy")
                                        .build())
                                .build(),
                        VideoBlock.builder()
                                .sortIndex(4)
                                .url("VideoBlockurl")
                                .type(VideoBlockType.YOUTUBE)
                                .build()
                ))).build();

        // when
        ArticleDto dto = articleMapper.articleToArticleDto(article);

        // then
        assertThat(dto).isEqualToIgnoringGivenFields(ArticleDto.builder()
                .author("author")
                .id(12345678L)
                .description("description")
                .title("title")
                .build(), "blocks");

        assertThat(dto.getBlocks())
                .extracting(ArticleBlockDto::getSortIndex)
                .isSorted();

        assertThat(dto.getBlocks()).containsExactlyInAnyOrder(
                ImageBlockDto.builder()
                        .sortIndex(1)
                        .image(ImageDto.builder()
                                .id(123L)
                                .imageSize(ImageSize.MEDIUM)
                                .url("other_url")
                                .build())
                        .build(),
                GalleryBlockDto.builder()
                        .sortIndex(2)
                        .images(asList(
                                ImageDto.builder()
                                        .id(123L)
                                        .imageSize(ImageSize.LARGE)
                                        .url("url")
                                        .build()
                        )).build(),
                TextBlockDto.builder()
                        .sortIndex(3)
                        .text("text")
                        .build(),
                VideoBlockDto.builder()
                        .sortIndex(4)
                        .url("VideoBlockurl")
                        .type(VideoBlockType.YOUTUBE)
                        .build()
        );
    }
}

File: Backend/mapping-kotlin/src/test/kotlin/com/mhp/coding/challenges/mapping/mappers/ArticleMapperTest.kt
package com.mhp.coding.challenges.mapping.mappers

import com.mhp.coding.challenges.mapping.models.db.Article
import com.mhp.coding.challenges.mapping.models.db.Image
import com.mhp.coding.challenges.mapping.models.db.ImageSize
import com.mhp.coding.challenges.mapping.models.db.blocks.*
import com.mhp.coding.challenges.mapping.models.dto.ArticleDto
import com.mhp.coding.challenges.mapping.models.dto.ImageDto
import com.mhp.coding.challenges.mapping.models.dto.blocks.GalleryBlockDto
import com.mhp.coding.challenges.mapping.models.dto.blocks.ImageBlockDto
import com.mhp.coding.challenges.mapping.models.dto.blocks.TextBlockDto
import com.mhp.coding.challenges.mapping.models.dto.blocks.VideoBlockDto
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.junit.jupiter.SpringExtension
import java.util.*

@SpringBootTest
@ExtendWith(SpringExtension::class)
class ArticleMapperTest(@Autowired private val mapper: ArticleMapper) {

    val fullArticle = Article(
        id = 1001,
        lastModified = Date(),
        lastModifiedBy = "Hans Müller",
        title = "Article Nr.: 1001",
        description = "Article Description 1001",
        author = "Max Mustermann",
        blocks = setOf(
            GalleryBlock(
                images = listOf(
                    Image(
                        id = 2,
                        url = "https://someurl.com/image/2",
                        imageSize = ImageSize.LARGE,
                        lastModified = Date(),
                        lastModifiedBy = "John Doe"
                    )
                ),
                sortIndex = 0
            ),
            TextBlock(
                text = "Some Text for 1",
                sortIndex = 0
            ),
            ImageBlock(
                image = Image(
                    id = 1,
                    url = "https://someurl.com/image/1",
                    imageSize = ImageSize.LARGE,
                    lastModified = Date(),
                    lastModifiedBy = "John Doe"
                ),
                sortIndex = 1
            ),
            VideoBlock(
                type = VideoBlockType.YOUTUBE,
                url = "https://youtu.be/myvideo",
                sortIndex = 4
            )
        ),
    )

    val expectedFullArticleDto = ArticleDto(
        id = 1001,
        title = "Article Nr.: 1001",
        description = "Article Description 1001",
        author = "Max Mustermann",
        blocks = listOf(
            GalleryBlockDto(
                images = listOf(
                    ImageDto(
                        id = 2,
                        url = "https://someurl.com/image/2",
                        imageSize = ImageSize.LARGE
                    )
                ),
                sortIndex = 0
            ),
            TextBlockDto(
                text = "Some Text for 1",
                sortIndex = 0
            ),
            ImageBlockDto(
                image = ImageDto(
                    id = 1,
                    url = "https://someurl.com/image/1",
                    imageSize = ImageSize.LARGE
                ),
                sortIndex = 1
            ),
            VideoBlockDto(
                type = VideoBlockType.YOUTUBE,
                url = "https://youtu.be/myvideo",
                sortIndex = 4
            )
        ),
    )

    @Test
    fun `mapper should map full article`() {
        val result = mapper.mapToDto(fullArticle)
        Assertions.assertThat(result).isEqualTo(expectedFullArticleDto)
    }

    val minimalArticle = Article(
        id = 1001,
        lastModified = Date(),
        title = "Article Nr.: 1001",
        blocks = setOf(
            GalleryBlock(
                images = listOf(
                    null,
                    Image(
                        id = 3,
                        url = "https://someurl.com/image/3",
                        imageSize = ImageSize.LARGE,
                        lastModified = Date(),
                    )
                ),
                sortIndex = 0
            ),
            ImageBlock(
                sortIndex = 1,
                image = null,
            ),
            VideoBlock(
                type = VideoBlockType.YOUTUBE,
                url = "https://youtu.be/myvideo",
                sortIndex = 4
            )
        ),
    )

    val expectedMinimalArticleDto = ArticleDto(
        id = 1001,
        title = "Article Nr.: 1001",
        description = "",
        author = "",
        blocks = listOf(
            GalleryBlockDto(
                images = listOf(
                    ImageDto(
                        id = 3,
                        url = "https://someurl.com/image/3",
                        imageSize = ImageSize.LARGE
                    )
                ),
                sortIndex = 0
            ),
            ImageBlockDto(
                image = ImageDto(
                    id = -1,
                    url = "",
                    imageSize = ImageSize.SMALL
                ),
                sortIndex = 1
            ),
            VideoBlockDto(
                type = VideoBlockType.YOUTUBE,
                url = "https://youtu.be/myvideo",
                sortIndex = 4
            )
        ),
    )

    @Test
    fun `mapper should map minimal article`() {
        val result = mapper.mapToDto(minimalArticle)
        Assertions.assertThat(result).isEqualTo(expectedMinimalArticleDto)
    }
}
